scripts/dotnet-cli-build/PrepareTargets.cs

```diff
@@ -200,6 +172,162 @@ public static BuildTargetResult RestorePackages(BuildTargetContext c)
             return c.Success();
         }
 
+        [Target]
+        public static BuildTargetResult CheckPrereqDebianPackageBuildComponents(BuildTargetContext c)
+        {
+            if (!CurrentPlatform.IsUbuntu)
+            {
+                return c.Success();
+            }
+
+            var debianPackageBuildDependencies = new string[]
+            {
+                "devscripts", 
+                "debhelper",
+                "build-essential"
+            };
+
+            var messageBuilder = new StringBuilder();
+
+            foreach (var package in debianPackageBuildDependencies)
+            {
+                if (!AptPackageIsInstalled(package))
+                {
+                    messageBuilder.Append($"Error: Debian package build dependency {package} missing.");
+                    messageBuilder.Append(Environment.NewLine);
+                    messageBuilder.Append($"-> install with apt-get install {package}");
+                    messageBuilder.Append(Environment.NewLine);
+                }
+            }
+
+            if (messageBuilder.Length == 0)
+            {
+                return c.Success();
+            }
+            else
+            {
+                return c.Failed(messageBuilder.ToString());
+            }
+        }
+
+        [Target]
+        public static BuildTargetResult CheckPrereqCoreclrDependencyPackages(BuildTargetContext c)
+        {
+            if (!CurrentPlatform.IsUbuntu && !CurrentPlatform.IsCentOS)
```

Comment:
@Sridhar-MS and I have been brainstorming on this. 

We've toyed with ideas of Target Dependency groups where a target is chosen based on the best fitting sets of constraints. This seems to complicate the overall problem without adding much additional benefit in our opinion.

We've also thought about Targets having a set of conditional actions which it will blindly run expecting the actions themselves to handle their conditions. This seems to just add another unnecessary layer of complexity to the target model.

Ultimately we've come to two suggestions which seem to fit the problem best. A developer will need to construct his target dependency graph through the use of dependencies the way it is today. Given a parent target like so:
```
[Target(CheckWindowsCoreclrDependencies, 
    CheckUbuntuCoreclrDependencies, 
    CheckCentOSCoreclrDependencies, 
    CheckOSXCoreclrDependencies, ...)]
public static BuildTargetResult CheckCoreclrPlatformDependencies(BuildTargetContext c) => c.Success();
```

The child items could be implemented in one of the two following ways. 

The first as suggested by @piotrpMSFT gives ultimate flexibility. It provides a `[TargetConditions(params Func<bool>)]` attribute. A nested example:
```
[Target(CheckWindowsX86CoreclrDependencies, CheckWindowsX64CoreclrDependencies)]
[TargetConditions(CurrentPlatform.IsWindows)
public static BuildTargetResult CheckWindowsCoreclrDependencies(BuildTargetContext c) => c.Success();

[Target()]
[TargetConditions(() => ProcessorArch == "x86"))
public static BuildTargetResult CheckWindowsX86CoreclrDependencies(BuildTargetContext c)
{
    // do things
}
```

The second suggestion takes a similar idea but makes the conditions themselves more explicit through individual Attributes. In other words every class of conditions will have it's own attribute to specify the parameters for that condition. Another example:

```
[Target(CheckWindowsX86CoreclrDependencies, CheckWindowsX64CoreclrDependencies)]
[Platform(PlatformEnum.Windows)] // This would be a params PlatformEnum
public static BuildTargetResult CheckWindowsCoreclrDependencies(BuildTargetContext c) => c.Success();

[Target()]
[Architecture("x86"))
public static BuildTargetResult CheckWindowsX86CoreclrDependencies(BuildTargetContext c)
{
    // do things
}
```

In both models all dependencies of a target will have their conditions evaluated before being run. If the conditions are not met, the target will be skipped, printing out some message but not throwing an exception.

We're curious to get everyone's thoughts, have a discussion and close on this today if possible.

cc @anurse How would this fit with your vision for the build framework?